In [3]:
import pandas as pd
import numpy as np
import os

In [87]:
os.getcwd()

'/Users/rahuljain/Desktop/Python/kaggle/titanic/input'

In [9]:
os.listdir()

['test.csv', 'train.csv', 'gender_submission.csv']

In [8]:
os.chdir('/Users/rahuljain/Desktop/Python/kaggle/titanic/input')

In [74]:
df = pd.read_csv('train.csv')

In [75]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Technique 1 - Check the percentage of missing values -- 
Perc missing > 95% - Drop the column
              95 < % Missing < 50 - No imputation required.
              % Missing < 50 - impute

In [76]:
desc = df.describe().T # transpose the describe dataframe and check for count column

In [77]:
desc

count        mean         std   min       25%       50%    75%  \
PassengerId  891.0  446.000000  257.353842  1.00  223.5000  446.0000  668.5   
Survived     891.0    0.383838    0.486592  0.00    0.0000    0.0000    1.0   
Pclass       891.0    2.308642    0.836071  1.00    2.0000    3.0000    3.0   
Age          714.0   29.699118   14.526497  0.42   20.1250   28.0000   38.0   
SibSp        891.0    0.523008    1.102743  0.00    0.0000    0.0000    1.0   
Parch        891.0    0.381594    0.806057  0.00    0.0000    0.0000    0.0   
Fare         891.0   32.204208   49.693429  0.00    7.9104   14.4542   31.0   

                  max  
PassengerId  891.0000  
Survived       1.0000  
Pclass         3.0000  
Age           80.0000  
SibSp          8.0000  
Parch          6.0000  
Fare         512.3292

In [78]:
##generate another column which will check the columns having missing %
desc['%Missing'] = 1 - ( desc['count'])/len(df)

In [79]:
desc

count        mean         std   min       25%       50%    75%  \
PassengerId  891.0  446.000000  257.353842  1.00  223.5000  446.0000  668.5   
Survived     891.0    0.383838    0.486592  0.00    0.0000    0.0000    1.0   
Pclass       891.0    2.308642    0.836071  1.00    2.0000    3.0000    3.0   
Age          714.0   29.699118   14.526497  0.42   20.1250   28.0000   38.0   
SibSp        891.0    0.523008    1.102743  0.00    0.0000    0.0000    1.0   
Parch        891.0    0.381594    0.806057  0.00    0.0000    0.0000    0.0   
Fare         891.0   32.204208   49.693429  0.00    7.9104   14.4542   31.0   

                  max  %Missing  
PassengerId  891.0000  0.000000  
Survived       1.0000  0.000000  
Pclass         3.0000  0.000000  
Age           80.0000  0.198653  
SibSp          8.0000  0.000000  
Parch          6.0000  0.000000  
Fare         512.3292  0.000000

In [80]:
drop_column_with_high_missing_val = []
for index,missing in desc['%Missing'].items():
    if missing > .95:
        drop_column_with_high_missing_val.append(index)
        

In [81]:
drop_column_with_high_missing_val

[]

In [82]:
df.drop(drop_column_with_high_missing_val,axis=1,inplace=True)

In [83]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [84]:
desc

count        mean         std   min       25%       50%    75%  \
PassengerId  891.0  446.000000  257.353842  1.00  223.5000  446.0000  668.5   
Survived     891.0    0.383838    0.486592  0.00    0.0000    0.0000    1.0   
Pclass       891.0    2.308642    0.836071  1.00    2.0000    3.0000    3.0   
Age          714.0   29.699118   14.526497  0.42   20.1250   28.0000   38.0   
SibSp        891.0    0.523008    1.102743  0.00    0.0000    0.0000    1.0   
Parch        891.0    0.381594    0.806057  0.00    0.0000    0.0000    0.0   
Fare         891.0   32.204208   49.693429  0.00    7.9104   14.4542   31.0   

                  max  %Missing  
PassengerId  891.0000  0.000000  
Survived       1.0000  0.000000  
Pclass         3.0000  0.000000  
Age           80.0000  0.198653  
SibSp          8.0000  0.000000  
Parch          6.0000  0.000000  
Fare         512.3292  0.000000

# Technique 2 - Check the standard deviation of all the columns  -- 
Very high std > drop them else keep
Also drop features with 0 std.. as contact values


# Technique 3 - Check the correlation coefficient within preictor variables:
    If corr cof is high... check the one predictor which has higher corr with target var and keep it
    drop the other 

In [85]:
df.corr()

PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000 -0.005007 -0.035144  0.036847 -0.057527 -0.001652   
Survived       -0.005007  1.000000 -0.338481 -0.077221 -0.035322  0.081629   
Pclass         -0.035144 -0.338481  1.000000 -0.369226  0.083081  0.018443   
Age             0.036847 -0.077221 -0.369226  1.000000 -0.308247 -0.189119   
SibSp          -0.057527 -0.035322  0.083081 -0.308247  1.000000  0.414838   
Parch          -0.001652  0.081629  0.018443 -0.189119  0.414838  1.000000   
Fare            0.012658  0.257307 -0.549500  0.096067  0.159651  0.216225   

                 Fare  
PassengerId  0.012658  
Survived     0.257307  
Pclass      -0.549500  
Age          0.096067  
SibSp        0.159651  
Parch        0.216225  
Fare         1.000000

Just for example Pclass and Fare are highly correlated (-0.549500) and now Pclass is better correlated with Survived than Fare.
-0.338481 vs 0.257307.. so we can drop Fare

Generally we can keep the Corr cof tolerance as .65

# Technique 4 - Check the multicollinearity- Remove these using PCA
if two or more variables are highly correlated then drop all but one from them

# Technique 5 - Check the multicollinearity- Remove using Cluster Analysis - Somewhtat similar to PCA
if two or more variables are highly correlated then drop all but one from them

# Technique 6 - Forward , Backward and Stepwise selection methog
We can use Mallow's CP as a criteria as well